02 - Variables, types, control flow.ipynb

```diff
@@ -0,0 +1,2206 @@
+   "execution_count": 18,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "5\n",
+      "H\n",
+      "===H===\n",
+      "===e===\n",
+      "===l===\n",
+      "===l===\n",
+      "===o===\n"
+     ]
+    }
+   ],
+   "source": [
+    "s = \"Hello\"\n",
+    "print(len(s))  # the length\n",
+    "print(s[0])    # the first character (element @ index 0)\n",
+    "\n",
+    "for char in s:\n",
+    "    print(\"===\" + char + \"===\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Обработката на низове става сравнително лесно чрез предоставените ни вградени функции и методи:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Deutschland ueber alles!\n",
+      "(nofilter) (nomakeup) (felt_cute_might_delete_later) (f4f) (follow4follow)\n",
+      "Access granted.\n"
+     ]
+    }
+   ],
+   "source": [
+    "with_umlauts = \"Deutschland über alles!\"\n",
+    "no_umlauts = with_umlauts.replace(\"ü\", \"ue\")  # returns a new string in which all occurrences of \"ü\" are replaced with \"ue\"\n",
+    "print(no_umlauts)\n",
+    "\n",
+    "post_hastags = \"#nofilter #nomakeup #felt_cute_might_delete_later #f4f #follow4follow\"\n",
+    "tags = post_hastags.replace(\"#\", \"\").split(\" \")  # first removes all #'s, then splits using a space as a delimiter into a list of strings\n",
+    "marked_tags_list = \" \".join(f\"({tag})\" for tag in tags)  # surrounds each element of the list with brackets and concatenates them with spaces\n",
+    "print(marked_tags_list)\n",
+    "\n",
+    "allowed_inputs = [\"y\", \"yes\", \"proceed\", \"correct\", \"da\", \"yy\", \"yesh\", \"yass\"]\n",
+    "user_input = \"Yes\"  # hardcoded for simplicity\n",
+    "if user_input.lower() in allowed_inputs:\n",
+    "    print(\"Access granted.\")\n",
+    "else:\n",
+    "    print(\"Access denied.\")\n",
+    "\n",
+    "# .lower() converts the string to lowercase, .upper() converts it to uppercase\n",
+    "# there is also .capitalize() which capitalizes only the first letter of the string"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Булеви\n",
+    "\n",
+    "Тип `bool` има две възможни стойности: `True` и `False`.\n",
+    "\n",
+    "За разлика от повечето други езици, вместо `!`, `||` и `&&` използваме `not`, `or` и `and`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "False\n",
+      "True\n",
+      "False\n",
+      "True\n"
+     ]
+    }
+   ],
+   "source": [
+    "t = True\n",
+    "f = not t\n",
+    "\n",
+    "print(f)\n",
+    "print(t or f)\n",
+    "print(t and f)\n",
+    "print(t and not f)"
```

Comment:
I agree that it would be easier to follow if we print an explanation too and I will change it. Not sure about the asserts though, maybe in some places they can help visualize things better but most of the time it would seem like an alien concept to most students I believe

Reply:
I agree with Alex on this one - let's not introduce `assert` in the second lecture